Synthetic _source: support ignore_above

nik9000 opened the pull request:
This allows you to use `ignore_above` with `keyword` fields in synthetic
source. Ignored values are stored in a "backup" stored field and added
to the end of the list of results. This makes `ignore_above` work pretty
much the same way as it does when you don't have synthetic source. The
only difference is the order of the results. But synthetic source
changes the order of results anyway. That should be fine.

nik9000 committed:
Synthetic _source: support ignore_above

This allows you to use `ignore_above` with `keyword` fields in synthetic
source. Ignored values are stored in a "backup" stored field and added
to the end of the list of results. This makes `ignore_above` work pretty
much the same way as it does when you don't have synthetic source. The
only difference is the order of the results. But synthetic source
changes the order of results anyway. That should be fine.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @nik9000, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

nik9000 committed:
Update docs/changelog/89466.yaml

nik9000 commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/400_synthetic_source.yml:
```diff
@@ -29,6 +29,7 @@ keyword:
           query:
             ids:
               values: [1]
+  - is_false: hits.hits.0.fields
```
There's a different bug here that I'm fixing along side `ignore_above` but I only noticed it when I was working on this. Previously we were adding all stored fields we loaded to the fetched field list. Now we only add if you ask for them.

nik9000 committed:
Cleaner?

nik9000 commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -257,7 +257,9 @@ private static FieldsVisitor createStoredFieldsVisitor(
-                    return new CustomFieldsVisitor(sourceLoader.requiredStoredFields(), true);
+                    // add the stored fields needed to load the source mapping to an empty set so they aren't returned
+                    sourceLoader.requiredStoredFields().forEach(fieldName -> storedToRequestedFields.putIfAbsent(fieldName, Set.of()));
+                    return new CustomFieldsVisitor(storedToRequestedFields.keySet(), true);
```
This change fixes that leaking `fields` in the search response. It's tricky!

romseygeek replied:
Sneaky.  Nice catch.

nik9000 commented on server/src/main/java/org/elasticsearch/index/fieldvisitor/FieldsVisitor.java:
```diff
@@ -84,6 +84,9 @@ public Set<String> getFieldNames() {
             MappedFieldType fieldType = fieldTypeLookup.apply(entry.getKey());
+            if (fieldType == null) {
+                continue; // TODO this is lame
+            }
```
I don't like adding such looseness. I'm sure there is a better way than this, but I couldn't think of something off the bat and figured we'd be refactoring here before too long anyway. So I wanted to see what other folks think.

I don't particularly like that this `valueForDisplay` thing is mixed up into the Lucene loading. That feels like it should be something we take care of externally but I have no idea how mixed up it is with other things.

romseygeek replied:
Yeah this really should be pulled out elsewhere, but I think I'm fine with it staying in here for the time being.

nik9000 replied:
That's what I was thinking. I don't want us to start relying on this null-ok behavior in other places. I don't particularly want to build some fancy abstraction for this either. Any ideas on something a little more defensive than this but not super overkill if we're going to rip it out?

nik9000 commented:
Note for those following along at home - this doesn't really effect the performance of loading synthetic source any - even if there are many, many fields, all of which have `ignore_above`. It's in the microseconds to build the extra HashMap entries. If you are worrying about microseconds you want to keep the mapping small.

nik9000 committed:
Fixup changelog

nik9000 committed:
More error message

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -945,6 +959,10 @@ private void indexValue(DocumentParserContext context, String value) {
             context.addIgnoredField(name());
+            if (context.isSyntheticSource()) {
+                // Save a copy of the field so synthetic source can load it
+                context.doc().add(new Field(originalName(), value, ORIGINAL_FIELD_TYPE));
```
If `stored=true` is set on here then we store it twice, right? Is it worth detecting that case and deferring to the 'normal' stored field rather than always using the hidden field?

nik9000 replied:
If I did that then fields longer than the limit would start being loadable via stored fields. That feels like something folks could accidentally rely on.

romseygeek reviewed (commented):
Thanks, I think this makes sense.  I have a question around what to do if the field is already being stored.

nik9000 committed:
More tests for enrich processor

Adds more tests for the enrich processor around different index types.
Right now they all work fine (yay!) but this feels like a good amount of
paranoia.

nik9000 committed:
Merge branch 'main' into synthetic_source_ignore_above_3

nik9000 committed:
Rename

nik9000 committed:
Merge branch 'main' into synthetic_source_ignore_above_3

nik9000 committed:
Merge branch 'main' into synthetic_source_ignore_above_3

nik9000 committed:
Update after merge

nik9000 commented:
Ready for you again @romseygeek !

nik9000 committed:
Done?

romseygeek reviewed (approved):
LGTM

nik9000 commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/get/100_synthetic_source.yml:
```diff
@@ -33,6 +33,7 @@ keyword:
   - match:
       _source:
         kwd: foo
+#  - is_false: fields  TODO fix me
```
It's the same problem as we have on fetch! I just noticed it and didn't want to forget it. I thought maybe I'd fix it like I did the fetch one. But it was a little more complex so I didn't. But I'll grab it soon!
